DESIGN.md

```diff
@@ -0,0 +1,205 @@
+- Universal Address
+    A 32-byte address, used by the wormhole contracts
+- Native Address (I think this should be called "Platform Address")
+    An address in the standard chain-specific format
+- Native
+    The "home" chain (e.g. ETH is native to Ethereum)
+- Foreign
+    A non-native chain (e.g. ETH is foreign to Moonbeam)
+- VAA (Verified Action Approval)
+    The core messaging primitive in Wormhole, it contains information about a message and a payload encoded as bytes.  Once finality is achieved and it is observed by the majority of the guardians, it is signed and can then be used to complete a transfer on the destination chain
+- Payload
+    Bytes that can be passed along with any wormhole message that contain application-specific data
+- Finality/Finality Threshold
+    The required number of blocks to wait until a VAA is produced
+
+# Discussion
+
+
+## What's the purpose of the Wormhole class?
+
+Wormhole class provides the main interface to do _everything_
+
+- Registers Platforms to access later -- constructor
+- Provides access to PlatformContexts -- getContext(ChainName)
+- Provides "shortcuts" to start a WormholeTransfer -- tokenTransfer/nftTransfer/cctp/...
+- Helpers for getting VAAs? or generally querying the API?
+- Abstract away chain-specific logic for easy mode access to methods
+
+## What do we want from a PlatformContext and how is that different from a provider / common utilities for a given platform?
+
+Provides Platform specific logic for a set of things
+
+- Register Modules (contract/app specific functionality)
+- Translates Platform specific stuff to generic stuff (e.g. ethers.Provider => RPC connection)
+- Deals with Platform specific interaction w/ chain (approve on eth, postvaa on sol, ...)
+- Implements standardized method format
+
+## What's the relationship between platforms/chains/providers?
+
+- A Platform provides the logic for all chains that run on that platform
+- A Chain provides consts (rpc/contract addresses/chain specific overrides)
+- A Provider is just an RPC connection and is held by the Chain. Providers are an implementation detail.
+
+## What's a signer vs. a wallet? Should signers have a provider (my answer: no)?
+
+- A Signer is an interface to sign transactions
+- It _may_ be backed by a wallet but not necessarily, as long as it fulfils the interface
+
+## Can we provide some way to make other non-standard applications available to use through the WormholeTransfer?
+
+Say I have an app that defines its own protocol, can I provide something that adheres to the WormholeTransfer interface so a dev can install it and call it like the TokenTransfer?
+
+
+# Outstanding Questions: 
+
+What is the preferred terminology to refer to either end of a cross-chain message: from/to, source/target or origin/destination?
+
+What is the preferred terminology for the core Wormhole layer? (i.e. Core Contracts or Wormhole Contracts)
+
+Should we namespace export base/definitions? 
```

Comment:
We shouldn't export the payloads by default because they only clutter up the namespace unnecessarily. E.g. users don't care about governance payloads and so they shouldn't be added to the payload factory by default.

Being a C++ dinosaur and fan-boy I think we should follow the same principles "you don't pay for things you don't use".